File: docs/RPC/jsonrpc.js
// SPDX-License-Identifier: CC0-1.0
"use strict";

/** Implements methods for accessing an Anope JSON-RPC server. */
class AnopeRPC {

	/**
	 * Initializes a new AnopeRPC instance with the specified RPC host.
	 *
	 * @param {string} host The RPC host base URL.
	 * @param {string} token The bearer token for authorizing with the RPC interface.
	 */
	constructor(host, token = "") {
		this.host = host;
		this.token = token;
	}

	/**
	 * Executes an arbitrary RPC query.
	 *
	 * @param {string} method The name of the method to execute.
	 * @param {...*} params The parameters pass to the method.
	 * @returns {*} The result of the RPC query.
	 */
	async run(method, ...params) {
		const body = JSON.stringify({
			"jsonrpc": "2.0",
			"method": method,
			"params": params.map((p) => p.toString()),
			"id": Math.random().toString(36).slice(2)
		});
		const headers = new Headers();
		if (this.token) {
			headers.append("Authorization", `Bearer ${btoa(this.token)}`);
		}
		const response = await fetch(this.host, {
			method: 'POST',
			headers: headers,
			body: body
		});
		if (!response.ok) {
			throw new Error(`HTTP returned ${response.status}`)
		}
		const json = await response.json();
		if ("error" in json) {
			throw new Error(`JSON-RPC returned ${json.error.code}: ${json.error.message}`)
		}
		if ("result" in json) {
			return json.result;
		}
		return null;
	}

	/**
	 * Retrieves a list of accounts.
	 *
	 * Requires the rpc_data module to be loaded.
	 *
	 * @param {string} detail The level of detail to request.
	 * @returns {(array|object)} A list of accounts.
	 */
	listAccounts(detail = "name") {
		return this.run("anope.listAccounts", detail);
	}

	/**
	 * Retrieves information about the specified account.
	 *
	 * Requires the rpc_data module to be loaded.
	 *
	 * @param {string} name The name of the account.
	 * @returns {object} An object containing information about the account.
	 */
	account(name) {
		return this.run("anope.account", name);
	}

	/**
	 * Retrieves a list of channels.
	 *
	 * Requires the rpc_data module to be loaded.
	 *
	 * @param {string} detail The level of detail to request.
	 * @returns {(array|object)} A list of channels.
	 */
	listChannels(detail = "name") {
		return this.run("anope.listChannels", detail);
	}

	/**
	 * Retrieves information about the specified channel.
	 *
	 * Requires the rpc_data module to be loaded.
	 *
	 * @param {string} name The name of the channel.
	 * @returns {object} An object containing information about the channel.
	 */
	channel(name) {
		return this.run("anope.channel", name);
	}

	/**
	 * Retrieves a list of services operators.
	 *
	 * Requires the rpc_data module to be loaded.
	 *
	 * @param {string} detail The level of detail to request.
	 * @returns {(array|object)} A list of services operators.
	 */
	listOpers(detail = "name") {
		return this.run("anope.listOpers", detail);
	}

	/**
	 * Retrieves information about the specified services operator.
	 *
	 * Requires the rpc_data module to be loaded.
	 *
	 * @param {string} name The name of the services operator.
	 * @returns {object} An object containing information about the services operator.
	 */
	oper(name) {
		return this.run("anope.oper", name);
	}

	/**
	 * Retrieves a list of servers.
	 *
	 * Requires the rpc_data module to be loaded.
	 *
	 * @param {string} detail The level of detail to request.
	 * @returns {(array|object)} A list of servers.
	 */
	listServers(detail = "name") {
		return this.run("anope.listServers", detail);
	}

	/**
	 * Retrieves information about the specified server.
	 *
	 * Requires the rpc_data module to be loaded.
	 *
	 * @param {string} name The name of the server.
	 * @returns {object} An object containing information about the server.
	 */
	server(name) {
		return this.run("anope.server", name);
	}

	/**
	 * Retrieves a list of users.
	 *
	 * Requires the rpc_data module to be loaded.
	 *
	 * @param {string} detail The level of detail to request.
	 * @returns {(array|object)} A list of users.
	 */
	listUsers(detail = "name") {
		return this.run("anope.listUsers", detail);
	}

	/**
	 * Retrieves information about the specified user.
	 *
	 * Requires the rpc_data module to be loaded.
	 *
	 * @param {string} nick The nick of the user.
	 * @returns {object} An object containing information about the user.
	 */
	user(nick) {
		return this.run("anope.user", nick);
	}

	/**
	 * Sends a message to every user on the network.
	 *
	 * Requires the rpc_message module to be loaded.
	 *
	 * @param {...*} messages One or more messages to send.
	 */
	messageNetwork(...messages) {
		return this.run("anope.messageNetwork", ...messages);
	}

	/**
	 * Sends a message to every user on the specified server.
	 *
	 * Requires the rpc_message module to be loaded.
	 *
	 * @param {string} name The name of the server.
	 * @param {...*} messages One or more messages to send.
	 */
	messageServer(server, ...messages) {
		return this.run("anope.messageServer", server, ...messages);
	}

	/**
	 * Sends a message to the specified user.
	 *
	 * Requires the rpc_message module to be loaded.
	 *
	 * @param {string} source The source pseudoclient to send the message from.
	 * @param {string} target The target user to send the message to.
	 * @param {...*} messages One or more messages to send.
	 */
	messageUser(source, target, ...messages) {
		return this.run("anope.messageUser", source, target, ...messages);
	}

	/**
	 * Checks whether the specified credentials are valid.
	 *
	 * Requires the rpc_user module to be loaded.
	 *
	 * @param {string} account A nickname belonging to the account to check.
	 * @param {string} password  The password for the specified account.
	 * @returns {object} An object containing basic information about the account.
	 */
	checkCredentials(account, password) {
		return this.run("anope.checkCredentials", account, password);
	}

	/**
	 * Identifies an IRC user to the specified account.
	 *
	 * Requires the rpc_user module to be loaded.
	 *
	 * @param {string} account Either an account identifier or nickname belonging to the account to
	 *                         identify to.
	 * @param {string} password The nickname of the IRC user to identify to the account.
	 */
	identify(account, user) {
		return this.run("anope.identify", account, user);
	}

	/**
	 * Lists all commands that exist on the network.
	 *
	 * Requires the rpc_user module to be loaded.
	 *
	 * @param {...*} services The nicknames of the services to list commands for.
	 * @returns {object} An object containing information about the available commands.
	 */
	listCommands(...services) {
		return this.run("anope.listCommands", ...services);
	}

	/**
	 * Lists all commands that exist on the network.
	 *
	 * Requires the rpc_user module to be loaded.
	 *
	 * @param {string} account If non-empty then the account to execute the command as.
	 * @param {string} service The service which the command exists on.
	 * @param {...*} command The the command to execute and any parameters to pass to it.
	 * @returns {object} An object containing information about the available commands.
	 */
	command(account, service, ...command) {
		return this.run("anope.command", account, service, ...command);
	}
}

/*
const arpc = new AnopeRPC("http://127.0.0.1:8080/jsonrpc");
arpc.listServers("full").then(servers => {
	console.log(servers);
}).catch (error => {
	console.log(error);
});
*/

File: docs/RPC/jsonrpc.rb
# SPDX-License-Identifier: CC0-1.0

require "base64"
require "json"
require "net/http"

# Implements methods for accessing an Anope JSON-RPC server.
class AnopeRPC

	# The RPC host base URL.
	attr_accessor :host

	# The bearer token for authorizing with the RPC interface.
	attr_accessor :token

	# Initializes a new AnopeRPC instance with the specified RPC host.
	#
	# @param host [String] The RPC host base URL.
	# @param token [String] The bearer token for authorizing with the RPC interface.
	def initialize(host, token = nil)
		@host = URI(host)
		@token = token
	end

	# Executes an arbitrary RPC query.
	#
	# @param method [String] The name of the method to execute.
	# @param params [Object] The parameters pass to the method.
	# @return [Object] The result of the RPC query.
	private def run(method, *params)
		body = {
			jsonrpc: "2.0",
			method: method,
			params: params.map(&:to_s),
			id: rand(36**16).to_s(36)
		}.to_json

		headers = {}
		headers["Authorization"] = "Bearer #{Base64.strict_encode64(@token)}" if @token

		response = Net::HTTP.post(@host, body, headers)
		raise "HTTP returned #{response.status}" unless response.is_a?(Net::HTTPSuccess)

		json = JSON.parse(response.body, symbolize_names: true)
		raise "JSON-RPC returned #{json[:error][:code]}: #{json[:error][:message]}" if json.key?(:error)

		json.dig(:result)
	end

	# Retrieves a list of accounts.
	#
	# Requires the rpc_data module to be loaded.
	#
	# @param detail [Symbol] detail The level of detail to request.
	# @return [Array] If the detail level is set to :name then an array of account names.
	# @return [Hash] If the detail level is set to :full then a hash of information about the
	#                accounts.
	def list_accounts(detail = :name)
		self.run("anope.listAccounts", detail)
	end

	# Retrieves information about the specified account.
	#
	# Requires the rpc_data module to be loaded.
	#
	# @param name [String] The name of the account.
	# @return [Hash] A hash containing information about the account.
	def account(name)
		self.run("anope.account", name)
	end

	# Retrieves a list of channels.
	#
	# Requires the rpc_data module to be loaded.
	#
	# @param detail [Symbol] detail The level of detail to request.
	# @return [Array] If the detail level is set to :name then an array of channel names.
	# @return [Hash] If the detail level is set to :full then a hash of information about the
	#                channels.
	def list_channels(detail = :name)
		self.run("anope.listChannels", detail)
	end

	# Retrieves information about the specified channel.
	#
	# Requires the rpc_data module to be loaded.
	#
	# @param name [String] The name of the channel.
	# @return [Hash] A hash containing information about the channel.
	def channel(name)
		self.run("anope.channel", name)
	end

	# Retrieves a list of services operators.
	#
	# Requires the rpc_data module to be loaded.
	#
	# @param detail [Symbol] detail The level of detail to request.
	# @return [Array] If the detail level is set to :name then an array of services operator names.
	# @return [Hash] If the detail level is set to :full then a hash of information about the
	#                services operators.
	def list_opers(detail = :name)
		self.run("anope.listOpers", detail)
	end

	# Retrieves information about the specified services operator.
	#
	# Requires the rpc_data module to be loaded.
	#
	# @param name [String] The name of the services operator.
	# @return [Hash] A hash containing information about the services operator.
	def oper(name)
		self.run("anope.oper", name)
	end

	# Retrieves a list of servers.
	#
	# Requires the rpc_data module to be loaded.
	#
	# @param detail [Symbol] detail The level of detail to request.
	# @return [Array] If the detail level is set to :name then an array of server names.
	# @return [Hash] If the detail level is set to :full then a hash of information about the
	#                servers.
	def list_servers(detail = :name)
		self.run("anope.listServers", detail)
	end

	# Retrieves information about the specified server.
	#
	# Requires the rpc_data module to be loaded.
	#
	# @param name [String] The name of the server.
	# @return [Hash] A hash containing information about the server.
	def server(name)
		self.run("anope.server", name)
	end

	# Retrieves a list of users.
	#
	# Requires the rpc_data module to be loaded.
	#
	# @param detail [Symbol] detail The level of detail to request.
	# @return [Array] If the detail level is set to :name then an array of user nicknames.
	# @return [Hash] If the detail level is set to :full then a hash of information about the
	#                users.
	def list_users(detail = :name)
		self.run("anope.listUsers", detail)
	end

	# Retrieves information about the specified user.
	#
	# Requires the rpc_data module to be loaded.
	#
	# @param name [String] The name of the user.
	# @return [Hash] A hash containing information about the user.
	def user(name)
		self.run("anope.user", name)
	end

	# Sends a message to every user on the network.
	#
	# Requires the rpc_message module to be loaded.
	#
	# @param messages [Array<String>] One or more messages to send.
	def message_network(*messages)
		self.run("anope.messageNetwork", *messages)
	end

	# Sends a message to every user on the specified server.
	#
	# Requires the rpc_message module to be loaded.
	#
	# @param name [String] The name of the server.
	# @param messages [Array<String>] One or more messages to send.
	def message_server(server, *messages)
		self.run("anope.messageServer", server, *messages)
	end

	# Sends a message to the specified user.
	#
	# Requires the rpc_message module to be loaded.
	#
	# @param source [String] The source pseudoclient to send the message from.
	# @param target [String] The target user to send the message to.
	# @param messages [Array<String>] One or more messages to send.
	def message_user(source, target, *messages)
		self.run("anope.messageUser", source, target, *messages)
	end
end

=begin
arpc = AnopeRPC.new("http://127.0.0.1:8080/jsonrpc")
begin
	pp arpc.list_servers(:full)
rescue StandardError => err
	STDERR.puts err
end
=end
